src/borrow_tracker/tree_borrows/diagnostics.rs

```diff
@@ -21,6 +38,8 @@ pub struct Event {
     pub access_kind: AccessKind,
     /// Relative position of the tag to the one used for the access
     pub is_foreign: bool,
+    /// Whether this access was explicit or inserted implicitly by Tree Borrows.
+    pub access_cause: AccessCause,
```

Comment:
I think I would prefer to leave the diagnostic changes to another PR, this one is getting too big... also then the diff in that PR makes it much easier to evaluate the change.